homeassistant/components/device_tracker/owntracks.py

```diff
@@ -60,47 +67,101 @@ def owntracks_event_update(topic, payload, qos):
 
         elif data['event'] == 'leave':
-            location = STATE_NOT_HOME
+            current_location = hass.states.get(
+                "device_tracker.{}".format(dev_id)).state
+
+            regions = REGIONS_ENTERED[dev_id]
+            if location in regions:
+                regions.remove(location)
+            new_region = regions[-1] if regions else None
+
+            if not new_region:
+                _LOGGER.info("Exit from %s to GPS", location)
+                see(**kwargs)
+
+                # if gps location didn't set new state, force to away
+                current_location = hass.states.get(
+                    "device_tracker.{}".format(dev_id)).state
+
+                if current_location.lower() == location.lower():
+                    kwargs['location_name'] = STATE_NOT_HOME
+                    see(**kwargs)
+
+                return
+
+            if current_location.lower() == new_region.lower():
+                _LOGGER.info("Exit from %s, still in %s",
+                             location, current_location)
+                return
+
+            zone = hass.states.get("zone.{}".format(new_region))
+            kwargs['location_name'] = new_region
+            _get_gps_from_zone(kwargs, zone)
+            _LOGGER.info("Exit from %s to %s", location, new_region)
+            see(**kwargs)
+
         else:
-            logging.getLogger(__name__).error(
+            _LOGGER.error(
                 'Misformatted mqtt msgs, _type=transition, event=%s',
                 data['event'])
             return
 
-        parts = topic.split('/')
-        kwargs = {
-            'dev_id': '{}_{}'.format(parts[1], parts[2]),
-            'host_name': parts[1],
-            'gps': (data['lat'], data['lon']),
-            'location_name': location,
-        }
-        if 'acc' in data:
-            kwargs['gps_accuracy'] = data['acc']
+    mqtt.subscribe(hass, LOCATION_TOPIC, owntracks_location_update, 1)
 
-        see(**kwargs)
+    mqtt.subscribe(hass, EVENT_TOPIC, owntracks_event_update, 1)
 
-    use_events = config.get(CONF_TRANSITION_EVENTS)
+    return True
 
-    if use_events:
-        mqtt.subscribe(hass, EVENT_TOPIC, owntracks_event_update, 1)
-    else:
-        mqtt.subscribe(hass, LOCATION_TOPIC, owntracks_location_update, 1)
 
-    return True
+def _parse_see_args(topic, data):
+    parts = topic.split('/')
+    dev_id = '{}_{}'.format(parts[1], parts[2])
+    host_name = parts[1]
+    kwargs = {
+        'dev_id': dev_id,
+        'host_name': host_name,
+        'gps': (data['lat'], data['lon'])
+    }
+    if 'acc' in data:
+        kwargs['gps_accuracy'] = data['acc']
+    if 'batt' in data:
+        kwargs['battery'] = data['batt']
+    return dev_id, kwargs
+
+
+def _get_gps_from_zone(kwargs, zone):
+    if zone is not None:
+        kwargs['gps'] = (
+            zone.attributes['latitude'],
+            zone.attributes['longitude'])
+        kwargs['gps_accuracy'] = 1
```

Comment:
This should be the accuracy of the zone, because we know the person is somewhere in that region.